File: src/app/runtime.ts
import { existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { TsqError } from "../errors";

export const DEFAULT_SNAPSHOT_EVERY = 100;

export function nowIso(): string {
  return new Date().toISOString();
}

/** Walk up from cwd to find the nearest directory containing `.tasque/`. */
export function findTasqueRoot(): string | null {
  let dir = process.cwd();
  for (;;) {
    if (existsSync(join(dir, ".tasque"))) {
      return dir;
    }
    const parent = dirname(dir);
    if (parent === dir) {
      return null;
    }
    dir = parent;
  }
}

/**
 * Return the project root containing `.tasque/`.
 * Falls back to cwd when no `.tasque/` is found (needed for `tsq init`).
 */
export function getRepoRoot(): string {
  return findTasqueRoot() ?? process.cwd();
}

export function getActor(repoRoot: string): string {
  const fromEnv = process.env.TSQ_ACTOR?.trim();
  if (fromEnv) {
    return fromEnv;
  }

  const gitName = readGitUserName(repoRoot);
  if (gitName) {
    return gitName;
  }

  const osUser = process.env.USERNAME ?? process.env.USER;
  if (osUser && osUser.trim().length > 0) {
    return osUser.trim();
  }

  return "unknown";
}

export function parsePriority(raw: string | number): 0 | 1 | 2 | 3 {
  const value = typeof raw === "number" ? raw : Number.parseInt(raw, 10);
  if (value === 0 || value === 1 || value === 2 || value === 3) {
    return value;
  }
  throw new TsqError("VALIDATION_ERROR", "priority must be one of: 0, 1, 2, 3", 1);
}

export function normalizeStatus(raw: string) {
  const normalized = raw === "done" ? "closed" : raw === "todo" ? "open" : raw;
  if (
    normalized === "open" ||
    normalized === "in_progress" ||
    normalized === "blocked" ||
    normalized === "closed" ||
    normalized === "canceled" ||
    normalized === "deferred"
  ) {
    return normalized;
  }
  throw new TsqError(
    "VALIDATION_ERROR",
    "status must be one of: open, todo, in_progress, blocked, closed, done, canceled, deferred",
    1,
  );
}

function readGitUserName(repoRoot: string): string | null {
  try {
    const proc = Bun.spawnSync({
      cmd: ["git", "config", "user.name"],
      cwd: repoRoot,
      stdout: "pipe",
      stderr: "ignore",
    });
    if (proc.exitCode !== 0) {
      return null;
    }
    const value = Buffer.from(proc.stdout).toString("utf8").trim();
    return value.length > 0 ? value : null;
  } catch {
    return null;
  }
}

File: src/app/runtime.rs
use crate::errors::TsqError;
use crate::types::{Priority, TaskStatus};
use chrono::{SecondsFormat, Utc};
use std::path::{Path, PathBuf};
use std::process::Command;

pub const DEFAULT_SNAPSHOT_EVERY: usize = 100;

pub fn now_iso() -> String {
    Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true)
}

pub fn find_tasque_root() -> Option<PathBuf> {
    let mut dir = std::env::current_dir().ok()?;
    loop {
        if dir.join(".tasque").exists() {
            return Some(dir);
        }
        let parent = dir.parent()?.to_path_buf();
        if parent == dir {
            return None;
        }
        dir = parent;
    }
}

pub fn get_repo_root() -> PathBuf {
    find_tasque_root()
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")))
}

pub fn get_actor(repo_root: impl AsRef<Path>) -> String {
    if let Ok(value) = std::env::var("TSQ_ACTOR") {
        let trimmed = value.trim();
        if !trimmed.is_empty() {
            return trimmed.to_string();
        }
    }

    if let Some(name) = read_git_user_name(repo_root) {
        return name;
    }

    let os_user = std::env::var("USERNAME")
        .or_else(|_| std::env::var("USER"))
        .ok();
    if let Some(value) = os_user {
        let trimmed = value.trim();
        if !trimmed.is_empty() {
            return trimmed.to_string();
        }
    }

    "unknown".to_string()
}

pub fn parse_priority(raw: &str) -> Result<Priority, TsqError> {
    let value = raw
        .parse::<u8>()
        .map_err(|_| TsqError::new("VALIDATION_ERROR", "priority must be one of: 0, 1, 2, 3", 1))?;
    if matches!(value, 0..=3) {
        return Ok(value);
    }
    Err(TsqError::new(
        "VALIDATION_ERROR",
        "priority must be one of: 0, 1, 2, 3",
        1,
    ))
}

pub fn normalize_status(raw: &str) -> Result<TaskStatus, TsqError> {
    let normalized = match raw {
        "done" => "closed",
        "todo" => "open",
        _ => raw,
    };
    match normalized {
        "open" => Ok(TaskStatus::Open),
        "in_progress" => Ok(TaskStatus::InProgress),
        "blocked" => Ok(TaskStatus::Blocked),
        "closed" => Ok(TaskStatus::Closed),
        "canceled" => Ok(TaskStatus::Canceled),
        "deferred" => Ok(TaskStatus::Deferred),
        _ => Err(TsqError::new(
            "VALIDATION_ERROR",
            "status must be one of: open, todo, in_progress, blocked, closed, done, canceled, deferred",
            1,
        )),
    }
}

fn read_git_user_name(repo_root: impl AsRef<Path>) -> Option<String> {
    let output = Command::new("git")
        .args(["config", "user.name"])
        .current_dir(repo_root)
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    let value = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if value.is_empty() { None } else { Some(value) }
}
